src/number_theory/divisors.lean

```diff
@@ -0,0 +1,188 @@
+    refine ⟨h.1, λ hm, _⟩,
+    rw hm at h,
+    tauto },
+  { by_cases heq : n = m.succ, {simp [heq]},
+    right,
+    exact ⟨h.1, lt_of_le_of_ne (nat.le_of_dvd (nat.succ_pos _) h.1) heq⟩ }
+end
+
+@[simp]
+lemma mem_divisors_antidiagonal {m : ℕ × ℕ} :
+  m ∈ divisors_antidiagonal n ↔ m.fst * m.snd = n ∧ n ≠ 0 :=
+begin
+  simp only [divisors_antidiagonal, finset.Ico.mem, ne.def, finset.mem_filter, finset.mem_product],
+  rw and_comm,
+  apply and_congr_right,
+  rintro rfl,
+  split; intro h,
+  { rintro h0, linarith },
+  { repeat {rw [nat.add_one_le_iff, nat.pos_iff_ne_zero, ne.def, nat.lt_add_one_iff] },
+    rw mul_eq_zero at h, rw decidable.not_or_iff_and_not at h,
+    refine ⟨⟨h.1, _⟩, ⟨h.2, _⟩⟩,
+    exact nat.le_mul_of_pos_right (nat.pos_of_ne_zero h.2),
+    exact nat.le_mul_of_pos_left (nat.pos_of_ne_zero h.1), }
+end
+
+variable {n}
+
+lemma divisor_le {m : ℕ}:
+n ∈ divisors m → n ≤ m :=
+begin
+  rw divisors, simp only [and_imp, finset.Ico.mem, finset.mem_filter],
+  intros a b c, omega,
+end
+
+variable (n)
+
+@[simp]
+lemma divisors_zero : divisors 0 = ∅ := by { ext, simp }
+
+@[simp]
+lemma divisors_antidiagonal_zero : divisors_antidiagonal 0 = ∅ := by { ext, simp }
+
+@[simp]
+lemma divisors_antidiagonal_one : divisors_antidiagonal 1 = {(1,1)} :=
+by { ext, simp [nat.mul_eq_one_iff, prod.ext_iff], }
+
+lemma swap_mem_divisors_antidiagonal {n : ℕ} {x : ℕ × ℕ} (h : x ∈ divisors_antidiagonal n) :
+  x.swap ∈ divisors_antidiagonal n :=
+begin
+  rw [mem_divisors_antidiagonal, mul_comm] at h,
+  simpa,
+end
+
+lemma fst_mem_divisors_of_mem_antidiagonal {n : ℕ} {x : ℕ × ℕ} (h : x ∈ divisors_antidiagonal n) :
+  x.fst ∈ divisors n :=
+begin
+  rw mem_divisors_antidiagonal at h,
+  simp [dvd.intro _ h.1, h.2],
+end
+
+lemma snd_mem_divisors_of_mem_antidiagonal {n : ℕ} {x : ℕ × ℕ} (h : x ∈ divisors_antidiagonal n) :
+  x.snd ∈ divisors n :=
+begin
+  rw mem_divisors_antidiagonal at h,
+  simp [dvd.intro_left _ h.1, h.2],
+end
+
+@[simp]
+lemma map_swap_divisors_antidiagonal {n : ℕ} :
+  (divisors_antidiagonal n).map ⟨prod.swap, prod.swap_right_inverse.injective⟩
+  = divisors_antidiagonal n :=
+begin
+  ext,
+  simp only [exists_prop, mem_divisors_antidiagonal, finset.mem_map, function.embedding.coe_fn_mk,
+             ne.def, prod.swap_prod_mk, prod.exists],
+  split,
+  { rintros ⟨x, y, ⟨⟨rfl, h⟩, rfl⟩⟩,
+    simp [h, mul_comm] },
+  { rintros ⟨rfl, h⟩,
+    use [a.snd, a.fst],
+    simp [h, mul_comm] }
+end
+
+lemma sum_divisors_eq_sum_proper_divisors_add_self :
+∑ i in divisors n, i = ∑ i in proper_divisors n, i + n :=
+begin
+  by_cases n = 0,
+  { rw h,
+    have h1 : proper_divisors 0 = ∅, ext, rw proper_divisors, simp,
+    rw h1, simp },
+  { rw divisors_eq_proper_divisors_insert_self,
+    rw finset.sum_insert, rw add_comm,
+    apply not_proper_self, omega, }
+end
+
+/-- `n : ℕ` is perfect if and only the sum of the proper divisors of `n` is `n`. -/
+def perfect (n : ℕ) : Prop := ∑ i in proper_divisors n, i = n
+
```

Comment:
Could you add `theorem perfect_iff_sum_proper_divisors := rfl`, for symmetry with `perfect_iff_sum_divisors_eq_two_mul`?